AppCenterDataStorage/AppCenterDataStorage/Internal/Client/MSTokenExchange.m

```diff
@@ -3,43 +3,129 @@
 
 static NSString *const kMSPartitions = @"partitions";
+static NSString *const kMSStorageReadOnlyDbTokenKey = @"MSStorageReadOnlyDbToken";
+static NSString *const kMSStorageUserDbTokenKey = @"MSStorageUserDbToken";
 
 @implementation MSTokenExchange : NSObject
 
-+ (void)tokenAsync:(MSStorageIngestion *)httpClient
-           partitions:(NSArray *)partitions
++ (void)performDbTokenAsyncOperationWithHttpClient:(MSStorageIngestion *)httpClient
+           partition:(NSString *)partition
     completionHandler:(MSGetTokenAsyncCompletionHandler)completion {
+    
+    MSTokenResult *cachedToken = [MSTokenExchange retrieveCachedToken:partition];
+    
+    if(cachedToken == nil)
+    {   
+        // Payload.
+        NSError *jsonError;
+        NSData *payloadData = [NSJSONSerialization dataWithJSONObject:@{kMSPartitions : @[partition]} options:0 error:&jsonError];
+        
+        // Http call.
+        [httpClient sendAsync:payloadData
+            completionHandler:^(NSString *callId, NSHTTPURLResponse *response, NSData *data, NSError *error) {
+                MSLogVerbose([MSDataStore logTag], @"Get token callback, request Id %@ with status code: %lu", callId,
+                             (unsigned long)response.statusCode);
+                
+                // If completion is provided.
+                if (completion) {
+                    
+                    // Read tokens.
+                    NSError *tokenResponsejsonError;
+                    NSDictionary *jsonDictionary = [NSJSONSerialization JSONObjectWithData:data options:0 error:&tokenResponsejsonError];
+                    if (tokenResponsejsonError) {
+                        MSLogError([MSDataStore logTag], @"Can't deserialize tokens with error: %@", [tokenResponsejsonError description]);
+                        completion([[MSTokensResponse alloc] initWithTokens:nil], error);
+                    }
+                    
+                    [MSTokenExchange saveToken:[[MSTokenResult alloc] initWithDictionary:jsonDictionary[kMSTokens][0]]];
+                    // Create token result object.
+                    MSTokensResponse *tokens = [[MSTokensResponse alloc] initWithDictionary:jsonDictionary];
+                    completion(tokens, error);
+                }
+            }];
+    }
+    else
+    {
+        NSError *error;
+        completion([[MSTokensResponse alloc] initWithTokens:@[cachedToken]], error);
+    }
+}
+
++ (void)saveToken:(MSTokenResult *)tokenResult {
+    NSString *tokenString = [tokenResult serializeToString];
+    BOOL success = [MSKeychainUtil storeString:tokenString forKey:[MSTokenExchange tokenKeyNameForPartition:tokenResult.partition]];
+    if (success) {
+        MSLogDebug([MSDataStore logTag], @"Saved token in keychain for the partitionKey : %@.", tokenResult.partition);
+    } else {
+        MSLogWarning([MSDataStore logTag], @"Failed to save the token in keychain for the partitionKey : %@.", tokenResult.partition);
```

Comment:
What happens if it failed to save token? Seems the log level should be an error if it blocks any functionalities of data sync.